puree/src/main/java/com/mercari/puree/TagPattern.java

```diff
@@ -0,0 +1,96 @@
+package com.mercari.puree;
+
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+public class TagPattern {
+
+    private static final String SEPARATOR = "\\.";
+    private static final String ALL_WILD_CARD = "**";
+    private static final String WILD_CARD = "*";
+    private static final Pattern VALID_PATTERN_ALPHANUMERIC = Pattern.compile("^[A-Za-z0-9]+");
+
+    private String pattern;
+
+    public TagPattern() {
+        this.pattern = ALL_WILD_CARD;
+    }
+
+    private TagPattern(String pattern) {
+        this.pattern = pattern;
```

Comment:
Why not use isValidPattern in this constructor, to be sure only valid patterns are accepted?

Reply:
As talked offline, fromString is handling this case.